Public/Src/Engine/Scheduler/Scheduler.cs

```diff
@@ -3092,7 +3108,7 @@ private async Task<PipExecutionStep> ExecutePipStep(RunnablePip runnablePip)
 
             // If schedule is terminating (e.g., StopOnFirstFailure), cancel the pip
             // as long as (i) 'start' step has been executed, (ii) the pip is in running state, and (iii) the pip has not been cancelled before.
-            if (m_scheduleTerminating && step != PipExecutionStep.Start && GetPipRuntimeInfo(pipId).State == PipState.Running && !runnablePip.IsCancelled)
+            if (ShouldCancelPip(runnablePip))
             {
                 return runnablePip.Cancel();
             }

```

Comment:
I'd like to move this stuff to ExecuteAsyncOrEnqueue as we discussed; so that we will have one place for the cancellation-related transition.

Reply:
Tried that. The problem is that if I did that we'd have to duplicate the logic to transition a pip to the next state in ExecuteAsyncOrEnqueue(...) (so it will be there AND in ExecutePip(...)). Otherwise the original next state will still be the next one.

So I think the change I'm proposing here is cleaner: we just make sure that a cancelled pip reaches ExecutePipStep(...). Then ExecutePip(...) will make sure the pips gets transitioned to Cancel, and then, in the next cycle, the actual cancellation step will be executed.

---
In reply to: [283903593](https://github.com/microsoft/BuildXL/pull/300#discussion_r283903593) [](ancestors = 283903593)